File: SNetUnity/SNetServer.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.Net;
using System.Net.Sockets;
using System.Threading;
[System.Serializable]
public struct ClientData
{
    public string ip;
    public int port;
    public string name;
}
[RequireComponent(typeof(TaskManager))]
public class SNetServer : MonoBehaviour
{
    [SerializeField] int myPort=8080;
    [SerializeField] ClientData[] clientDatas;
    Dictionary<IPEndPoint, SNetClient> clients=new Dictionary<IPEndPoint, SNetClient>();
    public Dictionary<string, SNetClient> namedClients = new Dictionary<string, SNetClient>();
    IPEndPoint[] endPoints;
    Thread serverThread,statusThread;
    UdpClient socket;
    bool isWorking = false;
    int counter = 0;
    void Awake()
    {
        endPoints = new IPEndPoint[clientDatas.Length];
        for(int i = 0; i < clientDatas.Length; i++)
        {
            IPEndPoint ep= new IPEndPoint(IPAddress.Parse(clientDatas[i].ip), clientDatas[i].port);
            endPoints[i] = ep;
            clients.Add(ep, new SNetClient(ep));
            namedClients.Add(clientDatas[i].name, clients[ep]);
        }
        serverThread = new Thread(new ThreadStart(ServerRecvThread));
      //  statusThread = new Thread(new ThreadStart(ServerStatusThread));

        isWorking = true;
        socket = new UdpClient(myPort);
        serverThread.Start();
        //statusThread.Start();

    }
    void ServerRecvThread()
    {
        IPEndPoint sender=new IPEndPoint(IPAddress.Any,8080);
        while (isWorking)
        {
            byte[] bytes = socket.Receive(ref sender);
            if (clients.ContainsKey(sender))
            {
                clients[sender].ReciveBytes(bytes);
            }
            Thread.Sleep(2);
        }
    }
    //void ServerStatusThread()
    //{
    //    SPacket packet = new SPacket(SPacketId.STATUS);
    //    packet.Write((int)ClientStatus.CONNECTED);
    //    while (isWorking)
    //    {
    //        foreach(var ep in endPoints)
    //        {
    //            clients[ep].CheckStatus();
    //            SendToClient(ep, packet);
    //        }
    //        Thread.Sleep(200);
    //    }
    //}
    public void SendToClient(IPEndPoint ep,SPacket packet)
    {
        try
        {
            byte[] bytes = packet.ToBytes();
            socket.Send(bytes, bytes.Length, ep);
        }
        catch { }
    }
    public void SendToClient(string client, SPacket packet)
    {
        SendToClient(namedClients[client].endPoint,packet);
    }
    private void Update()
    {
        if (Input.GetKeyDown(KeyCode.Space))
        {
            SPacket packet = new SPacket(SPacketId.DATA);
            packet.Write((float)counter);
            SendToClient(endPoints[0], packet);
            counter += 1;
        }

    }
}

File: SNetPython/snetserver.py
import snetclient as sc
import spacket as sp
import socket
import struct
import threading
class SNetServer:
    def __init__(self, my_port=8080, ips=None, ports=None,names=None):
        if ips is None:
            ips = []
        if ports is None:
            ports = []
        if names is None:
            names = []
        self.my_port = my_port
        self.clients = {}
        self.named_clients={}
        self.server_thread = threading.Thread(target=self.server_recv_thread)
        self.is_working = False
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

        for i in range(len(ips)):
            endpoint = (ips[i], ports[i])
            self.clients[endpoint] = sc.SNetClient(endpoint)
            self.named_clients[names[i]] = self.clients[endpoint]

    def start(self):
        self.is_working = True
        self.sock.bind(('127.0.0.1', self.my_port))
        self.server_thread.start()

    def stop(self):
        self.is_working = False
        self.sock.close()
        self.server_thread.join()

    def server_recv_thread(self):
        while self.is_working:
            print(self.my_port)
            try:
                data, sender = self.sock.recvfrom(1024)
                if sender in self.clients:
                    self.clients[sender].receive_bytes(data)
            except:
                pass
            
            # Add any additional handling for unknown clients if needed
            # For example, you might want to create a new SNetClient for unknown senders.
            # self.clients[sender] = SNetClient(sender)

    def send_to_client(self, endpoint, packet):
        bytes_data = packet.to_bytes()
        try:
            if isinstance(endpoint,str):
                self.sock.sendto(bytes_data,self.named_clients[endpoint].endpoint)
            else:
                self.sock.sendto(bytes_data, endpoint)
        except:
            pass